### DISCUSSION TITLE
Possible to exclude `__main__` block when running `reportUndefinedVariable` rules?

### DISCUSSION THREAD

**Discussion by dalejung** [microsoft/pyright#5853] [Ideas]
I would find it useful if we could exclude variables defined in `__main__` when warning about undefined variables. Take the simple example:

```python
def howdy():
    print(hi)   # preferable use of `hi` would spit a warning


if __name__ == '__main__':
    hi = 1
    howdy()
```

This is valid python but only when run as a script and not when imported. IMO, it is cleaner if `__main__` variables can only affect `reportUndefinedVariable` within `__main__` itself.

It's easy enough just to rename `__main__` for a sec and see what warnings pop up, but having a flag to default this behavior would be great. 

My workflow is constantly throwing files from vim to a running ipython. Files usually start with a `__main__` block where logic gets refactored into importable functions. So I footgun myself with this on the regular.

**Comment by erictraut** [comment]
The logic in pyright for undefined variable detection uses code flow analysis to determine whether a symbol is potentially undefined. Code flow analysis works only within an execution scope (e.g. within a function or within the global scope of a module). It cannot be applied across execution scopes because it's generally impossible to determine statically the order in which execution scopes are executed, relative to each other.

I'll also note that any special casing for `__main__` would result in false positives in some cases, and we generally try to avoid false positives, even at the expense of some false negatives (errors that occur at runtime but are not reported at static analysis time).

It would also require a special case that I don't think is well motivated by a defensible principle.

For all of these reasons, it's unlikely we'd even implement what you're suggesting here.

**Reply by dalejung** [reply]
:+1: Makes sense. For some reason I thought `__main__` was already special cased.
